parameters.txt
<ECF>
    <Genotype>
        <Tree>
            <Entry key="functionset">+ - * / pos min max</Entry>
            <Entry key="terminalset">t1 t2 t3 t4 t5 t6 t7 t8 t9 t10 t11 t12 t13 t14 t15 t16</Entry>
            <Entry key="mindepth">1</Entry>
            <Entry key="maxdepth">7</Entry>
        </Tree>
    </Genotype>
    <Registry>
	    <Entry key="term.eval">10000</Entry>
        <Entry key="log.filename">log.txt</Entry>
        <Entry key="log.level">4</Entry>
    </Registry>
</ECF>
<Conf>
    <Main> <!-- Glavne opcije za rad programa -->
        <BestFile>best.txt</BestFile>
        <RunBest>0</RunBest> <!-- Izvrti jednu simulaciju koristeci jedinku iz BestFile -->
        <SeedSimulator>1</SeedSimulator> <!-- Ako 1 koristi se SimulatorSeed, inace nasumican seed -->
        <SimulatorSeed>4444</SimulatorSeed>
        <SimulationSteps>500</SimulationSteps> <!-- Broj koraka simulacije -->
        <PrintSteps>0</PrintSteps> <!-- Ispis svakog koraka u obliku teksta -->
        <MetaAlg>3</MetaAlg> <!-- Odabir meta-algoritma 0/1/2/3 -->
        <FeatureSelection>2 4 8 9 10 11 12 13 14 15 16 17 20 21 23 28</FeatureSelection>
        <StepDuration>400</StepDuration> <!-- Duljina koraka pri vizualizaciji (u ms) -->
    </Main>
    <KPI> <!-- KPI tezine -->
        <Weights>5 -4 1 -1 0 -10 0 -2</Weights>
    </KPI>
    <Simulator> <!-- Opcije koje mijenjaju rad simulatora -->
        <BufferCount>3</BufferCount>
        <MaxArrivalSize>3</MaxArrivalSize>
        <MaxBufferSize>12</MaxBufferSize>
        <MinInitialBuffer>3</MinInitialBuffer>
        <MaxInitialBuffer>5</MaxInitialBuffer>
        <MinDueTime>60</MinDueTime>
        <MaxDueTime>180</MaxDueTime>
        <MinWaitFactor>0.1</MinWaitFactor>
        <MaxWaitFactor>0.2</MaxWaitFactor>
        <ArrivalProbability>15</ArrivalProbability>
        <HandoverProbability>22</HandoverProbability>
        <InitalizeBuffers>1</InitalizeBuffers>
    </Simulator>
</Conf>

parameters_cgp.txt
<ECF>
	<Genotype>
		<Cartesian>
			<Entry key="numvariables">17</Entry>
			<Entry key="numoutputs">1</Entry>
			<Entry key="numrows">2</Entry>
			<Entry key="numcols">5</Entry>
			<Entry key="levelsback">3</Entry>
			<Entry key="functionset">+ * - / pos min max</Entry>
		</Cartesian>
	</Genotype>
    <Registry>
	    <Entry key="term.eval">10000</Entry>
        <Entry key="log.filename">log.txt</Entry>
        <Entry key="log.level">4</Entry>
    </Registry>
</ECF>
<Conf>
    <Main> <!-- Glavne opcije za rad programa -->
        <BestFile>best.txt</BestFile>
        <RunBest>0</RunBest> <!-- Izvrti jednu simulaciju koristeci jedinku iz BestFile -->
        <SeedSimulator>1</SeedSimulator> <!-- Ako 1 koristi se SimulatorSeed, inace nasumican seed -->
        <SimulatorSeed>4444</SimulatorSeed>
        <SimulationSteps>500</SimulationSteps> <!-- Broj koraka simulacije -->
        <PrintSteps>0</PrintSteps> <!-- Ispis svakog koraka u obliku teksta -->
        <MetaAlg>3</MetaAlg> <!-- Odabir meta-algoritma 0/1/2/3 -->
        <FeatureSelection>2 4 8 9 10 11 12 13 14 15 16 17 20 21 23 27</FeatureSelection>
        <StepDuration>400</StepDuration> <!-- Duljina koraka pri vizualizaciji (u ms) -->
    </Main>
    <KPI> <!-- KPI tezine -->
        <Weights>0.5 -0.4 0.1</Weights>
    </KPI>
    <Simulator> <!-- Opcije koje mijenjaju rad simulatora -->
        <BufferCount>3</BufferCount>
        <MaxArrivalSize>3</MaxArrivalSize>
        <MaxBufferSize>12</MaxBufferSize>
        <MinInitialBuffer>3</MinInitialBuffer>
        <MaxInitialBuffer>5</MaxInitialBuffer>
        <MinDueTime>60</MinDueTime>
        <MaxDueTime>180</MaxDueTime>
        <MinWaitFactor>0.1</MinWaitFactor>
        <MaxWaitFactor>0.2</MaxWaitFactor>
        <ArrivalProbability>15</ArrivalProbability>
        <HandoverProbability>22</HandoverProbability>
        <InitalizeBuffers>1</InitalizeBuffers>
    </Simulator>
</Conf>

trainer/src/Parameters.h
#ifndef PARAMETERS_H
#define PARAMETERS_H

#include <string>
#include <ECF/ECF.h>

class Parameters
{
private:
	template<typename T>
	static void readParam(XMLNode& conf, T& var, const char *parent, const char *name) {
		XMLNode par = conf.getChildNode(parent);
		if (par.isEmpty())
			return;

		using UnderlyingType = typename std::remove_const<typename std::remove_reference<T>::type>::type;

		if constexpr (std::is_same<UnderlyingType, std::string>::value)
			var = !par.getChildNode(name).isEmpty() ? par.getChildNode(name).getText() : var;
		else if constexpr (std::is_same<UnderlyingType, float>::value)
			var = !par.getChildNode(name).isEmpty() ? std::stof(par.getChildNode(name).getText()) : var;
		else
			var = !par.getChildNode(name).isEmpty() ? std::stoi(par.getChildNode(name).getText()) : var;
	}
public:
	// main
	inline static std::string BEST_FILE = "best.txt";
	inline static bool USING_ECF = false;
	inline static bool RUN_BEST = false;
	inline static int MODEL = -1; // 0 - GP; 1 - CGP
	// Runner
	inline static std::string KPI_WEIGHTS;
	inline static int STEP_DURATION = 400;
	// Simulator
	inline static bool PRINT_STEPS = false;
	inline static bool INITALIZE_BUFFERS = true;
	inline static int BUFFER_COUNT = 3;
	inline static int MAX_ARRIVAL_SIZE = 3;
	inline static int MAX_BUFFER_SIZE = 12;
	inline static int MIN_INIT_BUFFER = 3;
	inline static int MAX_INIT_BUFFER = 5;
	inline static int MIN_DUE_TIME = 70;
	inline static int MAX_DUE_TIME = 90;
	inline static float MIN_WAIT_FAC = 0.1;
	inline static float MAX_WAIT_FAC = 0.3;
	inline static int CRANE_TIME = 2;
	inline static int ARRIVAL_PROB = 15; // /100
	inline static int HANDOVER_PROB = 22; // /100
	inline static bool SEED_SIM = true; // seedaj sim prije pokretanja
	inline static int SIMULATOR_SEED = 4444; // pocetni seed
	// SimulatorEvalOp
	inline static int SIM_STEPS = 50;
	// PriorityHeuristic
	inline static int META_ALG = 0; // 0 default; 1,2,3 - kandidati
	inline static std::string SELECTED_FEATURES;

	static void readParameters(int argc, char** argv) {
		if (argc != 2) {
			std::cout << "Parameters file not provided properly!" << std::endl;
			std::cout << "Using default values!" << std::endl;
			return;
		}

		XMLNode params = XMLNode::parseFile(argv[1]);
		if (params.isEmpty()) {
			std::cout << "Parameters file (" << argv[1] << ") doesn't exist or isn't properly formated!" << std::endl;
			std::cout << "Using default values!" << std::endl;
			return;
		}

		XMLNode ecf = params.getChildNode("ECF");
		if (!ecf.isEmpty()) {
			USING_ECF = true;
			if (!ecf.getChildNodeByPath("Genotype/Tree").isEmpty())
				MODEL = 0;
			else if (!ecf.getChildNodeByPath("Genotype/Cartesian").isEmpty())
				MODEL = 1;
		}

		XMLNode conf = params.getChildNode("Conf");

		if (!conf.isEmpty()) {
			readParam(conf, RUN_BEST, "Main", "RunBest");
			readParam(conf, BEST_FILE, "Main", "BestFile");

			readParam(conf, KPI_WEIGHTS, "KPI", "Weights");
			readParam(conf, STEP_DURATION, "Main", "StepDuration");

			readParam(conf, PRINT_STEPS, "Main", "PrintSteps");
			readParam(conf, INITALIZE_BUFFERS, "Simulator", "InitalizeBuffers");
			readParam(conf, BUFFER_COUNT, "Simulator", "BufferCount");
			readParam(conf, MAX_ARRIVAL_SIZE, "Simulator", "MaxArrivalSize");
			readParam(conf, MAX_BUFFER_SIZE, "Simulator", "MaxBufferSize");
			readParam(conf, MIN_INIT_BUFFER, "Simulator", "MinInitialBuffer");
			readParam(conf, MAX_INIT_BUFFER, "Simulator", "MaxInitialBuffer");
			readParam(conf, MIN_DUE_TIME, "Simulator", "MinDueTime");
			readParam(conf, MAX_DUE_TIME, "Simulator", "MaxDueTime");
			readParam(conf, MIN_WAIT_FAC, "Simulator", "MinWaitFactor");
			readParam(conf, MAX_WAIT_FAC, "Simulator", "MaxWaitFactor");
			readParam(conf, ARRIVAL_PROB, "Simulator", "ArrivalProbability");
			readParam(conf, HANDOVER_PROB, "Simulator", "HandoverProbability");
			readParam(conf, CRANE_TIME, "Simulator", "CraneTime");
			readParam(conf, SEED_SIM, "Main", "SeedSimulator");
			readParam(conf, SIMULATOR_SEED, "Main", "SimulatorSeed");

			readParam(conf, SIM_STEPS, "Main", "SimulationSteps");

			readParam(conf, META_ALG, "Main", "MetaAlg");

			readParam(conf, SELECTED_FEATURES, "Main", "FeatureSelection");
		}
	}
};

#endif
